Initial processing of bindep.txt from collections

AlanCoding opened the pull request:
Connect #50

This still needs more cleanup, but I want people to know where I'm going.

This corresponds to a change to my `ovirt.ovrit` branch which adds the `bindep.txt` of:

```
gcc [platform:rpm]
python3-devel [platform:rpm]
libcurl-devel [platform:rpm]
openssl-devel [platform:rpm]
libxml2-devel [platform:rpm]
```

This produces the `Dockerfile`

```
FROM shanemcd/ansible-runner

RUN pip3 install --upgrade pip setuptools
ENV PYCURL_SSL_LIBRARY=openssl
ADD introspect.py /usr/local/bin/introspect
RUN chmod +x /usr/local/bin/introspect
ADD requirements.yml /build/

RUN ansible-galaxy role install -r /build/requirements.yml --roles-path /usr/share/ansible/roles
RUN ansible-galaxy collection install -r /build/requirements.yml --collections-path /usr/share/ansible/collections
RUN pip3 install bindep
RUN dnf -y install $(bindep -b -f /usr/share/ansible/collections/ansible_collections/ovirt/ovirt/bindep.txt)
ADD requirements.txt /build/
RUN pip3 install --upgrade -r /build/requirements.txt
```

For the awx case with the definition:

```
---
version: 1
dependencies:
  galaxy: requirements.yml
additional_build_steps:
  prepend:
    - RUN pip3 install --upgrade pip setuptools
    - ENV PYCURL_SSL_LIBRARY=openssl
```

And with that, it does run to completion with all the python requirements (particularly pycurl) installing.

AlanCoding committed:
Processing of bindep.txt from collections

Catch python requirements parsing errors in collections

Implementation of comments for sources
  inside bindep.txt and requirements.txt

update bindep steps contract and speculative existence check

Add integration test for bindep.txt from ansible.posix

Install bindep library oustide of Containerfile steps
Run bindep command outside of the Containerfile steps
  done as combined commands in podman run

AlanCoding commented on ansible_builder/cli.py:
```diff
@@ -25,17 +25,17 @@ def run():
-                print()
-                print('Dependency data for {0}'.format(folder))
-                print(yaml.dump(data, default_flow_style=False))
+        data = process(args.folder, user_pip=args.user_pip, user_bindep=args.user_bindep)
```
Changed `args.folders` to `args.folder` - because the original idea was that we would introspect all the default Ansible collection search paths. This is unlikely to ever be supported. It's also very hard to reconcile with combining with the user-provided files.

That's a major change - before, we combined the user's content with the collection content _after_ the introspect script. Now, I am moving the user's files into the image and having the introspect script include that content in its output.

AlanCoding commented on ansible_builder/introspect.py:
```diff
@@ -125,14 +163,76 @@ def get_dependency(self, entry):
+
+def parse_introspect_output(stdout):
+    p = re.compile(
+        '{0}(?P<yaml_text>.+){1}'.format(begin_delimiter, end_delimiter),
```
Added regex processing here.

Why:

Before, we called `docker run -rm <tag> introspect`. When we did that, the stdout constituted a full YAML document (usually).

Here, we changed `introspect` to be inside a `RUN` layer, as opposed to a command ran outside of the Containerfile. That means that its output is interlaced in with the rest of the build command output. Instead of running more unnecessary commands, I decided I would just use markers to extract it.

shanemcd replied:
I don't love this. What was the reasoning here?

AlanCoding replied:
Actually, https://github.com/ansible/ansible-builder/issues/78 factored into my thinking.

That issue wasn't important enough to act on, but then several other changes started getting thrown into the mix where `introspect` and `bindep` get thrown into a series of multiple commands. It became easier to assume that the output will get "padded" with other text, because it's not a hard problem to solve.

AlanCoding commented on ansible_builder/introspect.py:
```diff
@@ -125,14 +163,76 @@ def get_dependency(self, entry):
+    parser.add_argument(
+        '--user-bindep', dest='user_bindep',
+        help='An additional file to combine with collection bindep requirements.'
+    )
```
Again, this is changing it so that the introspect script combines the user's `requirements.txt` and `bindep.txt` files with those files inside of collections. So that gets added to the contract for the introspect command too.

```
$ ansible-builder introspect test/data/ --user-pip test/data/pip/project/requirements.txt 

Dependency data for test/data/
python:
  test.metadata:
  - pyvcloud>=14
  test.reqfile:
  - pytz
  - tacacs_plus
  - 'pyvcloud>=18.0.10  # duplicate with test.metadata collection'
  user:
  - bottle>=3.0
system:
  test.bindep:
  - subversion [platform:rpm]
  - subversion [platform:dpkg]
```

shows that it is picking up

```
$ cat test/data/pip/project/requirements.txt
bottle>=3.0
```

AlanCoding commented on ansible_builder/introspect.py:
```diff
@@ -125,14 +163,76 @@ def get_dependency(self, entry):
+    parser.add_argument(
+        '--write-bindep', dest='write_bindep',
+        help='Write the combined bindep file to this location.'
+    )
```
Not using `--write-pip`, but added it for consistency.

We have the introspect script write the bindep file because of technicalities of how bindep operates. The first phase needs to have written a file ready to pass to `bindep`

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -1,20 +1,25 @@
+}
+BINDEP_COMBINED = 'bindep_combined.txt'
+BINDEP_OUTPUT = 'bindep_output.txt'
+PIP_COMBINED = 'requirements_combined.txt'
```
I changed the naming of most things that go into the build context folder.

Before, we accepted the naming from the user.... I'd rather not, because these go through our own phases of combination, sanitization, and transformation. I want these to be as clear as possible so that we can document them.

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -49,18 +54,39 @@ def build_command(self):
+            self.container_runtime, 'run',
+            '--rm', self.tag, '/bin/bash', '-c'
+        ] + [' '.join(command)]
+        return run_command(wrapped_command, **kwargs)
```
I had to change this to the `-c` pattern so I could run more than 1 command with a single invocation.

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -49,18 +54,39 @@ def build_command(self):
+                [
+                    'pip3', 'install', 'bindep;',
+                    'echo', '----begin_bindep_output----;',
+                    'bindep', '-b', '-f', '/build/{0}'.format(BINDEP_COMBINED)],
```
Here, I both install and run bindep in a new container. This keeps bindep out of the python environment in the image, see https://github.com/ansible/ansible-builder/pull/76#issuecomment-674635447

shanemcd replied:
For downstream we'll need to have bindep in the base image, so we should probably do that upstream too.

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -191,21 +217,12 @@ def create_folder_copy_files(self):
-            if do_copy:
-                shutil.copy(requirement_path, self.build_context)
+            dest = os.path.join(self.build_context, new_name)
+            copy_file(requirement_path, dest)
```
Coding practices for copy/write had diverged in a few places, so I tried to fully finish shared methods for `copy_file` and `write_file`

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -228,43 +245,41 @@ def prepare_appended_steps(self):
-                user_py_reqs = f.read().split('\n')
-            data['python'].extend(user_py_reqs)
-        data['python'] = sanitize_requirements(data['python'])
-        pip_file = os.path.join(self.build_context, 'requirements.txt')
```
Combining user's `requirements.txt` file was moved earlier in the process (phase 1 as opposed to phase 2), so that's why it's axed here. This was a phase 2 thing.

AlanCoding commented on ansible_builder/requirements.py:
```diff
@@ -14,24 +14,28 @@
+                consolidated.append(req)
+                seen_pkgs.add(req.name)
+        except Exception as e:
+            print('Warning: failed to parse requirments from {}, error: {}'.format(collection, e))
```
Added try-except here so that we would always get a result as opposed to erring.

I don't think a single bad-actor collection should halt the entire process. If a collection has stuff that cannot be parsed, we don't want to include any of its requirements, because that's a red flag.

AlanCoding commented on ansible_builder/requirements.py:
```diff
@@ -40,11 +44,13 @@ def sanitize_requirements(py_reqs):
         else:
             raise RuntimeError('Could not process {0}'.format(req.line))
 
+        sanitized.append(new_line + '  # from collection {}'.format(','.join(req.collections)))
```
Added comments to denote the source of requirements.

AlanCoding committed:
Move awx EE bindep into ovirt.ovirt branch

AlanCoding committed:
Echo stdout in event of failures

shanemcd committed:
Rework how and when bindep is run

AlanCoding committed:
Remove introspect step

Add introspect method, move combine logic back to main
  Fill in missing file reads

fix line type bug

Split out introspect sanity tests and remove outdated param

further simplify ovirt.ovirt in awx example

Unit test copy_file just like write_file

shanemcd commented on ansible_builder/main.py:
```diff
@@ -49,18 +52,39 @@ def build_command(self):
+    def run_in_container(self, command, **kwargs):
+        wrapped_command = [self.container_runtime, 'run','--rm']
+
+        wrapped_command.extend(['-v', f"{os.path.abspath(self.build_context)}:/context:Z"])
```
If we're hardcoding this here, we can remove `volumes` from the kwargs passed to the invocations below.

AlanCoding replied:
Yes, did that. I wanted this to be an assumption about running in the container.

One more thing to note - `abspath` make the syntax more verbose

```
docker run --rm -v /Users/alancoding/Documents/repos/ansible-builder/test/data/bc:/context:Z exec-env-pycurl /bin/bash -c python3 /context/introspect.py
```

I don't know that it's necessary. It's clearly not broken, so I don't want to touch it now, just submitting it for consideration. It could be the case that it has `..` in it, so if we got rid of this, maybe we would have to resolve the relative path, which might take a little more legwork.

shanemcd commented on ansible_builder/main.py:
```diff
@@ -49,18 +52,39 @@ def build_command(self):
+
+        rc, output = self.run_in_container(
+            ['python3', '/context/introspect.py', '--write-bindep', '/context/bindep_combined.txt'],
+            volumes=[f"{os.path.abspath(self.build_context)}:/context:Z"],
```
This can go

shanemcd commented on ansible_builder/main.py:
```diff
@@ -49,18 +52,39 @@ def build_command(self):
+        if collection_data.get('system'):
+            rc, output = self.run_in_container(
+                ['bindep', '-b', '-f', '/context/{0}'.format(BINDEP_COMBINED)],
+                volumes=[f"{os.path.abspath(self.build_context)}:/context:Z"],
```
and this

AlanCoding commented:
I was eager to rebase, and wound up pushing this in a really half-implemented form

AlanCoding commented:
This should be ready for review after it is rebased after the merge of #88

AlanCoding commented:
Now this has been rebased, tests are passing, and hopefully all the dust has settled. Ready for review

AlanCoding commented on ansible_builder/main.py:
```diff
@@ -49,18 +52,55 @@ def build_command(self):
+            rc, bindep_output = self.run_in_container(
+                ['bindep', '-b', '-f', '/context/{0}'.format(BINDEP_COMBINED)],
+                allow_error=True, capture_output=True
+            )
```
hmm.... should we _delete_ the `BINDEP_COMBINED` file here? I learn towards "no" but feel like I have to ask the question.

AlanCoding commented:
regate

shanemcd commented:
regate

AlanCoding commented:
regate
